### PR TITLE
Proof of concept: Web Fonts API as an extension of WP_Dependencies

### LINKED ISSUE #41481: Webfonts API Architecture: Explore leveraging WP_Dependencies existing enqueuing, registration, and internal handling
Part of https://github.com/WordPress/gutenberg/issues/41479.

Explore changing the current architecture to extend from the `WP_Dependencies` to reuse its implementations of enqueuing, registration, and internal handling.

Why?

The current implementation re-implements a lot of the concepts that are already handled by `WP_Dependencies` for the script and style loaders. The script loader classes already have the concepts of dependencies, enqueuing, and registering, as well as a way to "do_item()" (generate CSS or a <link> tag with source, in this case).

Goals: reduce code, increase readability, and increase maintainability.

### PR DESCRIPTION
**Pull request by desrosj**
## What?
**DO NOT MERGE**

This is a proof of concept for the Web Fonts API as an extension of the `WP_Dependencies` class.

It doesn't work quite yet. This is mainly to gauge whether this would be a feasible option to pursue further.

## Why?
The current implementation re-implements a lot of the concepts that are already handled by `WP_Dependencies` for the script and style loaders. A `WP_Web_Fonts` class could potentially cut down on a lot of the code needed to implement a web fonts API.

The script loader classes already have the concepts of dependencies, enqueuing, and registering, as well as a way to "do_item()" (generate CSS or a `<link>` tag with source, in this case).

## How?
This uses the following:
- A font family is first registered. Behind the scenes, this registers an alias dependency (no script attached).
- Font variations are registered independently with their own unique handle (which allows extenders to remove specific font variations), and added as dependencies for the font family alias.
- Enqueuing the alias font family enqueues all registered variations for that family.
- Individual variations can be enqueued separately.

### CODE REVIEW

File: lib/experimental/class-wp-webfonts.php [hunk 1]
```diff
@@ -44,192 +12,177 @@ class WP_Webfonts {
...
-				_doing_it_wrong( __METHOD__, __( 'Could not determine the font family name.', 'gutenberg' ), '6.0.0' );
-				return false;
-			}
-		}
-
-		return sanitize_title( $to_convert );
+	public function add_dependency( $font_family_handle, $variation_handle ) {
+		$dependencies = $this->registered[ $font_family_handle ];
+		$dependencies[] = $variation_handle;
+		$this->registered[ $font_family_handle ] = $dependencies;
```
Comment (dream-encode) [inline lib/experimental/class-wp-webfonts.php]: Is there a reason this can't be simplified to this?

```suggestion
		$this->registered[ $font_family_handle ][] = $variation_handle;
```

Comment (desrosj) [reply lib/experimental/class-wp-webfonts.php]: Missed this initially! I actually had a typo here and was not properly referencing the dependencies correctly. I fixed the typo and applied this simplification.

**Comment by hellofromtonya** [comment]
Thank you @desrosj for creating this POC. A lot of the functionality that is here and currently in Gutenberg's implementation will be removed and/or changed as part of the [current vision in the ongoing roadmap](https://github.com/WordPress/gutenberg/issues/41479).

I have concerns about extending from `WP_Dependencies` to make it work for the needs of font management.

### tl;dr

1. Webfonts API does not need font-family to font-family dependency management which is big part of script/script loading management.
2. `WP_Dependencies` exposes functionality that will not be used in the Webfonts API.

I'm thinking Webfonts API doesn't really fit into the intent of `WP_Dependencies`.

### The concept of `dependency` is different

The understanding of "dependencies" for font management is different than styles/scripts.

`WP_Dependencies` and style/script loading management focuses more on managing stylesheet-to-stylesheet or script-to-script dependencies, i.e. one stylesheet is dependent upon one or other stylesheets or one script file is dependent upon one or more other script files.

Dependencies for styles and scripts have a 1:many relationship where it's clear the intent is to manage those resource-to-resource relationships to ensure a stylesheet or script files dependencies are also loaded into memory.

Font management is different. It doesn't have font-family to font-family dependencies. For example, `Source Serif Pro` is not dependent upon `Lato`.  A font is also not dependent upon other scripts or stylesheets.

### Unneeded `WP_Dependencies` public functionality

The following list of `WP_Dependencies` functionality would need to be disabled, meaning the publicly expose methods are still there but are not used or functional. 

IMO having the publicly exposed methods that aren't usable will be confusing to extenders and contributors.

#### Enqueue
Enqueuing will not be publicly exposed, but rather will be private inner-workings of the API (please refer to the [Vision > Enqueuing section](https://github.com/WordPress/gutenberg/issues/41479)).

`WP_Dependencies` publicly exposes the ability to enqueue and dequeue. Overloading could make the `enqueue()` and `dequeue` methods not do anything, ie empty methods.

#### Remove
Removing a font-family and/or a font-family's font-face (variant) are not planned functionality due to concerns of plugins removing theme defined font-families from registration.
 
Current vision:
* the theme sets the fonts
* plugins can also _register_ additional fonts, but cannot override theme defined fonts. Note: These registered fonts will be available for users to select through the Global Styles > Typography pickers.

`WP_Dependencies` publicly exposes the ability to remove a handle, which in this case would be a font-family. Overloading could make the `remove()` method not do anything, ie an empty method.


What do you think @desrosj?

**Comment by desrosj** [comment]
Closing this out as it has been replaced by other PRs (See above).
